_posts/2023-01-26-jpastreamer-extension.adoc

```diff
@@ -0,0 +1,164 @@
+    .sorted(Person$.firstName.comparator().thenComparing(Person$.lastName.comparator())) (2)
+    .skip(5) (3)
+    .limit(10) (4)
+    .collect(Collectors.toList())
+----
+<1> Filters can be combined with the and/or operators
+<2> Easily filter on one or more properties
+<3> Skip the first 5 Persons
+<4> Return at most 10 Persons
+
+In the context of queries, the Stream operators filter, sort, limit, and skip all have a natural mapping that makes the resulting query expressive and intuitive to read while remaining compact.
+
+The Stream above is translated by JPAStreamer to the following HQL statement: 
+[source, text]
+----
+select
+   person0_.person_id as person_id1_0_,
+   person0_.first_name as first_na2_0_,
+   person0_.last_name as last_nam3_0_,
+   person0_.created_at as created_4_0_,
+from
+   person person0_
+where
+   person0_.person_id=1
+where
+    (person0_.first_name like ?) 
+    and (person0_.last_name like ?)
+order by
+    person0_.first_name asc,
+    person0_.last_name asc limit ?, ?
+----
+
+== How JPAStreamer works
+Okay, it looks simple. But how does it work? JPAstreamer uses an annotation processor to form a meta-model at compile time. It inspects any classes marked with the standard JPA annotation `@Entity`, and for every entity `Foo.class`, a corresponding `Foo$.class` is created. The generated classes represent entity attributes as Fields used to form predicates on the form `User$.firstName.startsWith("A")` that can be interpreted by JPAStreamer's query optimizer.
+
+It is worth repeating that JPAStreamer does not alter or disturb the existing codebase but merely extends the API to handle Java Stream queries.
+
+== Installing the JPAstreamer Extension
+JPAStreamer is installed as any other Quarkus extension, using a Maven dependency:
+
+[source, xml]
+----
+<dependency>
+    <groupId>io.quarkiverse.jpastreamer</groupId>
+    <artifactId>quarkus-jpastreamer</artifactId>
+    <version>1.0.0</version>
+</dependency>
+----
+
+After the dependency is added, rebuild your Quarkus application to trigger JPAStreamer’s annotation processor. The installation is complete once the generated fields reside in `/target/generated-sources`; you’ll recognise them by the trailing $ in the classnames, e.g. `Person$.class`.
+
+NOTE: JPAStreamer requires an underlying JPA provider, such as Hibernate. For this reason, JPAStreamer needs no additional configuration as the database integration is taken care of by the JPA provider.
+
+== JPStreamer and Panache
+Any Panache fan will note that JPAStreamer shares some of its objectives with Panache, in simplifying many common queries. Still, JPAStreamer distinguishes itself by instilling more confidence in the queries with its type-safe Stream interface. Luckily however, no one if forced to take a pick as Panache and JPAStreamer work seamlessly alongside each other.
+
+NOTE: Here is an example Quarkus application that uses both JPAStreamer and Panache.
```

Comment:
Is a link to a repository missing?
I see the link is in the Resources below but you probably want it here too?

Reply:
You are right, thank you for pointing that out!

Reply:
Fixed in new commit.